File: Seminář Java (IJA)/Projekt/src/controls/Time.java
package controls;

import entities.Street;

/**
 * Time class holds information about simulated application time.
 * It also provides useful methods for work with simulated time and constructing time tables.
 * @author Michal Rein, Ondřej Motyčka
 */
public final class Time {

    /** Represents current simulated time in seconds. */
    long appTime = 0;
    /** Speed ratio, increasing this ration increases pace of time. */
    int speedRatio = 1;
    /** Delay counter. Delay is affected by speed ratio.*/
    int delayCounter = 0;
    /** Handle for status bar. */
    StatusBarTime timeStatusBarHandle;

    /**
     * Constructor of Time class. Sets current time to starting time given by parameter start.
     * @param start Starting value of time.
     * @param timeStatusBar Status bar handle.
     */
    public Time(int start, StatusBarTime timeStatusBar){
        appTime = start;
        timeStatusBarHandle = timeStatusBar;
    }

    /**
     * Simulate pass of one second in simulated world. After trespassing value of 86400 (which stands for 24 hours),
     * time must be reset. Status bar is also informed about time change and new value is displayed.
     */
    public void updateTime(){
        appTime++; // increment simulated application time
        if (appTime >= 86400) appTime = 0;  // after 24 hours, timer must be reset
        // update time on status bar
        timeStatusBarHandle.updateTime(convertTimeToString(appTime));
    }

    /**
     * Get current simulated time.
     * @return Simulated time (in seconds).
     */
    public long getAppTime(){
        return appTime;
    }

    /**
     * Reset of delay counter.
     */
    public void resetDelayCounter(){
        delayCounter = 0;
    }

    /**
     * Incrementation of delay counter.
     */
    public void incrementDelayCounter(){
        delayCounter++;
    }

    /**
     * Get delay counter.
     * @return Delay counter value.
     */
    public int getDelayCounter(){
        return delayCounter;
    }

    /**
     * Get speed ratio.
     * @return Speed ratio value.
     */
    public int getSpeedRatio(){
        return speedRatio;
    }

    /**
     * Set new speed ratio.
     * @param newSpeed New value of speed ratio.
     */
    public void setSpeedRation(int newSpeed){
        speedRatio = newSpeed;
    }

    /**
     * Method converts time given in seconds into HH:MM:SS format.
     * @param time Time which has to be converted into readable format.
     * @return Time in HH:MM:SS format.
     */
    public static String convertTimeToString(long time){
        long hours = time / 3600;
        long minutes = (time / 60) % 60;
        long secs = time % 60;
        return String.format("%02d", hours) + ":" + String.format("%02d", minutes) + ":" + String.format("%02d", secs);
    }

    /**
     * Method converts time given in HH:MM:SS format into seconds.
     * @param timeString Time in HH:MM:SS format.
     * @return Time in seconds.
     */
    public static long convertStringToTime(String timeString){
        long time = 0;
        String[] parts = timeString.split(":");
        time += Integer.parseInt(parts[0]) * 3600;
        time += Integer.parseInt(parts[1]) * 60;
        time += Integer.parseInt(parts[2]);
        return time;
    }

    /**
     * Calculate time difference between current time and given time.
     * @param time Time is seconds.
     * @return Difference of current time and time given by parameter.
     */
    public String calculateTimeDifference(long time){
        long newTime = time - appTime;
        return convertTimeToString(newTime);
    }

    /**
     * Return congestion level constant. It is used for calculation of estimated time arrival.
     * @param street Street from which we need to get congestion constant.
     * @return Congestion constant of given street.
     */
    public static int getCongestionConstant(Street street){
        int constant;
        switch (street.getCongestionLevel()){
            case 1:
                constant = 2;
                break;
            case 2:
                constant = 4;
                break;
            case 3:
                constant = 10;
                break;
            default:
                constant = 1;
                break;
        }
        return constant;
    }

}

File: Seminář C++ (ICP)/src/model/time.cpp
/*!
 * @file
 * @brief This file contains implementation of Time class
 *
 * @author Michal Rein, Ondřej Motyčka
*/

#include "time.h"

/**
 * @brief Time::Time Constructor of Time object.
 */
Time::Time()
{
    this->appTime = new QTime(0, 0, 0, 0);
}

/**
 * @brief Time::getAppTime Get actual QTime object.
 * @return QTime object of simulated time.
 */
QTime *Time::getAppTime()
{
    return this->appTime;
}

/**
 * @brief Time::getAppTimeInSeconds Extract iformation about simulated time and convert it into seconds.
 * @return Actual time in seconds.
 */
qint16 Time::getAppTimeInSeconds(){

    qint16 time = 0;
    time += this->appTime->hour() * 3600;
    time += this->appTime->minute() * 60;
    time += this->appTime->second();
    return time;
}

/**
 * @brief Time::updateTime Update simulated time, 1 second passed.
 */
void Time::updateTime()
{
    *this->appTime = appTime->addSecs(1);
}

/**
 * @brief Time::convertTimeToString Convert actual time to QString human readable format.
 * @param time Actual time in seconds.
 * @return Time string.
 */
QString Time::convertTimeToString(qint16 *time)
{
    qint16 hours = *time/3600;
    qint16 minutes = (*time/60) % 60;
    qint16 secs = *time % 60;
    QString timeString = QString("%1").arg(hours, 2, 10, QChar('0')) + ":" + QString("%1").arg(minutes, 2, 10, QChar('0')) + ":" + QString("%1").arg(secs, 2, 10, QChar('0'));
    return timeString;

}

/**
 * @brief Time::convertStringToTime Convert time string in HH:MM:SS format to seconds.
 * @param timeString Time string.
 * @return Time representing given time string in seconds.
 */
qint16 Time::convertStringToTime(QString *timeString)
{
    qint16 time = 0;
    QStringList parts = timeString->split(":");
    time += parts.value(0).toInt(nullptr, 10) * 3600;
    time += parts.value(1).toInt(nullptr, 10) * 60;
    time += parts.value(2).toInt(nullptr, 10);
    return time;
}

/**
 * @brief Time::getCongestionConstant Read congestion level of given street and return delay constant for correct timetable construction.
 * @param street Street object.
 * @return Congestion delay constant.
 */
int Time::getCongestionConstant(Street *street)
{
    int constant;
    switch (*street->getCongestion()) {
        case 1:
            constant = 2;
            break;
        case 2:
            constant = 4;
            break;
        case 3:
            constant = 10;
            break;
        default:
            constant = 1;
            break;
    }
    return constant;

}

